File: L1_insertion_sort.py
#Imports
from array import array
import random as r

# Constants
N = 10

# Globals
x = array('i', [0]*N)

# Subroutines

# Insertion Sort
def insertion_sort(A, n):
	"""Sort a list or numpy array.

	Argument:
	A -- a list or numpy array
	n -- length of A
	"""
	# Traverse the list or array from index 1 to n-1.
	for i in range(1, n):
		key = A[i]

		# Insert A[i] into the sorted subarray a[0:i].
		# Compare stored key with the already sorted values to its left.
		# Move each item one position to the right until we find the 
		# position for the key or fall off the left end of the list or array.
		j = i - 1
		while j >= 0 and A[j] > key:
			A[j + 1] = A[j]
			j -= 1

		# Insert key at the correct position in the list or array.
		A[j + 1] = key
# Printing
def print_list(list, n):
        for i in range(0, n):
                print("%3d" % (list[i]), end="")
        print()


# Main Routine
def main():

        # Initial header
        print("Lab1 - Insertion sort")
        print()
        
        # For list to initialize array elements
        for i in range (0, N):
                x[i] = r.randint(0, 99)

        # Print initialized elements and explanatory header
        print("List before sorting:")
        print_list(x, N)
        print()

        # Invoke insertion sort algorithm
        insertion_sort(x, N)

        # Print sorted list and explanatory header
        print("List after sorting:")
        print_list(x, N)
        
        
# Start-up Scripts
if __name__ == "__main__":
        main()

File: L1_insertion_sort.c
// Imports
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Constants
#define N 10

// Globals
int x[N];

// Subroutines


// Printing Routine
void print_list(int *list, int n)
{
  for(int i = 0; i<n; i++)
    {
      printf(" %3d", list[i]);
    };
  printf("\n");
}


// Insertion Sort Routine 
void insertion_sort(int *list, int n)
{
  int key;
  int j;
  
  for(int i = 1; i < N; i++)
    {
      key = list[i];
      j = i - 1;

      while((j>=0) && (list[j]>key))
	{
	  list[j +1] = list[j];
	  j -= 1;
	};
      list[j+1] = key;
    };
}

// Main routine
int main(void)
{
  srand(time(NULL));
  printf("Lab1 - Insertion Sort\n\n");

  for(int i = 0; i < N; i++)
    {
      x[i] = (rand()>>24);
    }

  printf("List before sorting\n");
  print_list(x, N);

  printf("\nList after sorting\n");
  insertion_sort(x, N);
  print_list(x, N);
  
  return 0;
}
